web3/contract/async_contract.py

```diff
@@ -307,6 +307,16 @@ async def call(
 
         block_id = await async_parse_block_identifier(self.w3, block_identifier)
 
+        # Proxy validation to make sure the contract function hits __call__()
+        if self.args is None and self.kwargs is None:
```

Comment:
I need clarification about this exception. In the original issue I understood that the values need to be initialized before calling the contract function. Does this prevent calling functions which do not have inputs?

I believe args and kwargs are both treated as optional inside of `async_call_contract_function` that encodes/validates the inputs. Maybe this also needs to consider validating against the ABI? That should be happening as part of the encoding step with my latest ABI updates. If no inputs are provided and they are defined in the ABI it should now throw a `MismatchedABI` (I think it used to throw the `TypeError`).

Here is the check for "encodability" with the arguments, and `TypeError`s are handled such that the calling function (`get_abi_element`) will raise a `MismatchedABI` with a [descriptive message](https://github.com/ethereum/web3.py/blob/90cb12c38383de0923bec667d643c7bed05b7e36/web3/utils/abi.py#L192) after the [filter](https://github.com/ethereum/web3.py/blob/90cb12c38383de0923bec667d643c7bed05b7e36/web3/utils/abi.py#L389) fails.

Reply:
> Does this prevent calling functions which do not have inputs?

Nope, the original fix added a new (better) error message when calling a function without parentheses, and without a contract instance. However, we've decided that it's more convenient to allow users to call ContractFunctions without parentheses if no arguments or keyword arguments are needed, so I've switched tactics here.